### PR TITLE
Patch

### PR DESCRIPTION
**Pull request by fourlastor**
## Problem

We need to wrap the [patch](https://developers.google.com/apis-explorer/#search/firestore/firestore/v1beta1/firestore.projects.databases.documents.patch) method to our api to apply validation and app authentication.

## Solution

This consists of 4 parts:
1. Apply authentication via middleware. App token is in `patch:app_token` inside firebase config.
2. Unwraps the original request to an object firestore can digest via js apis.
3. Apply validators (configurable, see code)
4. Save data to firebase.

#### HTTP Body

```json
{
  "fields": {
    "banana": {
      "stringValue": "banana"
    },
    "name": {
      "stringValue": "daniele"
    },
    "aString": {
      "stringValue": "abc"
    },
    "aBoolean": {
      "booleanValue": true
    },
    "anArray": {
      "arrayValue": {
        "values": [
          {
            "stringValue": "def"
          },
          {
            "stringValue": "123"
          }
        ]
      }
    },
    "anInteger": {
      "integerValue": 123
    },
    "aDouble": {
      "doubleValue": 123.4
    },
    "aMap": {
      "mapValue": {
        "fields": {
          "aString": {
            "stringValue": "abc"
          },
          "anInteger": {
            "integerValue": 123
          }
        }
      }
    },
    "aNull": {
      "nullValue": null
    },
    "aTimestamp": {
      "timestampValue": 1518902415827
    },
    "aReference": {
      "referenceValue": "/categories/1"
    }
  }
}
```

`http PATCH https://[firebase-project-domain].cloudfunctions.net/patch/bananas/1234 'Authorization: Bearer 123' < body.json`

#### Firestore object

```json
{
    "aBoolean": true,
    "aDouble": 123.4,
    "aMap": {
        "aString": "abc",
        "anInteger": 123
    },
    "aNull": null,
    "aReference": {
        "_referencePath": {
            "_databaseId": "(default)",
            "_formattedName": "projects/[firebase-project]/databases/(default)/documents/categories/1",
            "_projectId": "[firebase-project]",
            "segments": [
                "categories",
                "1"
            ]
        }
    },
    "aString": "abc",
    "aTimestamp": "2018-02-17T21:20:15.827Z",
    "anArray": [
        "def",
        "123"
    ],
    "anInteger": 123,
    "banana": "banana",
    "name": "daniele"
}
```

### CODE REVIEW

File: functions/src/patch/map-fields.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+} from './firestore-api'
+
+export const mapFields = (app: FirebaseApp) => (fields: Fields): {[key: string]: any} => {
+    return Object.keys(fields)
+        .reduce((result, key) => {
+            return { ...result, [key]: mapField(app)(fields[key]) }
+        }, {})
+}
+
+const mapField = (app: FirebaseApp) => (field: FieldValue): any => {
```
Comment (rock3r) [inline functions/src/patch/map-fields.ts]: If you only ever need the `firestore()` why not just passing that in?

File: functions/src/patch/patch.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
+import * as express from 'express'
+import { FirebaseApp } from '../firebase'
+import { mapFields } from './map-fields'
+import { Failure, required, Validator } from './validator'
+
+const patch = (firebaseApp: FirebaseApp) => {
+    const app = express()
```
Comment (rock3r) [inline functions/src/patch/patch.ts]: Is this variable naming not confusing? I wouldn't think `app` is an instance of `express`

Comment (fourlastor) [reply functions/src/patch/patch.ts]: Dunno, I think it's the standard (express naming the app instance app), maybe `expressApp`?

File: functions/src/patch/map-fields.ts (hunk 1 above)
Comment (fourlastor) [reply functions/src/patch/map-fields.ts]: Mostly because in every function I'm passing `FirebaseApp`

File: functions/src/patch/patch.ts (hunk 2 above)
Comment (rock3r) [reply functions/src/patch/patch.ts]: `expressApp` would be clearer for me, yeah :)

File: functions/src/patch/map-fields.ts (hunk 1 above)
Comment (rock3r) [reply functions/src/patch/map-fields.ts]: I see. I was thinking mostly in terms of minimising knowledge

**Pull request merged**